## ЧАСТЬ I. Решите задания используя только стандартные методы Python 3

### 1. Импортируйте число e из библиотеки math и выведите его с окргулением до 4 цифр после запятой

In [19]:
import math 
round(math.e, 4)

2.7183

### 2. Вычислите целую часть результата деления 100 на 7. Убедитесь, что её тип – целое число.

In [6]:
x = 100 // 7 
type(x)

int

### 3. Создайте список с квадратами целых чисел от 12 до 4 с шагом 2 не делящимися на 5.

In [22]:
x = [x*x for x in range(4, 13, 2) if x % 5 != 0 ]
x.reverse()
print(x)

[144, 64, 36, 16]


### 4.	Отсортируйте полученный в пункте 3 список относительно модуля разницы между элементами этого списка и числом 50.

In [38]:
x = [x*x for x in range(4, 13, 2) if x % 5 != 0 ]
print(x)
for i in range(len(x)):
    x[i] = abs(50 - x[i])
x.sort()
print(x)
    


[16, 36, 64, 144]
[14, 14, 34, 94]


### 5. Создайте множество (set) содержащее все 26 маленькие буквы латинского алфавита.

In [41]:
a = set('abcdefghijklmnopqrstuvwxyz')
print(a)

{'a', 'z', 'b', 'o', 'y', 'c', 'g', 'u', 'p', 'j', 'm', 'r', 's', 'v', 'd', 'h', 'f', 'l', 'w', 'k', 'n', 'i', 'q', 't', 'x', 'e'}


### 6.	Создайте список, содержащий числа с повторениями. Используя словарь, подсчитайте, сколько раз каждое число встречалось в этом списке, и напечатайте содержимое словаря

In [12]:
List = [2, 2, 4, 7, 3, 2, 10, 4, 3, 14]
Dict = {}

for i in List:
    if i in Dict:
        Dict[i] += 1
    else:
        Dict[i] = 1 

y

{2: 3, 4: 2, 7: 1, 3: 2, 10: 1, 14: 1}

### 7. Реализуйте функцию от трёх аргументов a, b и c находящую вещественные корни уравнения ax2 + bx + c = 0, с нулевыми значениями аргументов по умолчанию

In [70]:
import math

def fun(a = 0, b = 0, c = 0):
    d = b*b - 4*a*c
    if (d < 0):
        return 'Решения нет'
    elif (d > 0):
        x1 = (-b + sqrt(d)) / (2 * a)
        x2 = (-b - sqrt(d)) / (2 * a)
        print(x1)
        print(x2)
    else: print(-b / (2 * a))
    
fun(3, -18, 27)

3.0


### 8.	Реализуйте класс Point с полями x,y заполняемыми в конструкторе

In [4]:
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

SomePoint = Point(1,1)

### 9. Реализуйте класс Circle наследника класса Point с дополнительным полем r и методом inside(p) проверяющим, что круг одержит точку p.

In [20]:
class Circle(Point):
    def __init__(self, x, y, r):
        Point.__init__(self, x, y)
        self.r = r
    
    def inside(self, p):
        if (pow((p.x - self.x), 2) + pow((p.y - self.y), 2) <= pow(self.r, 2)):
            print(f'Точка {p.x, p.y} принадлежит кругу')
        else:
            print(f'Точка {p.x, p.y} не принадлежит кругу')
        
SomeCircle = Circle(3, 3, 3)
SomeCircle.inside(SomePoint)
x1 = Point(0, 0)
с1 = Circle(2, 0, 2)
с1.inside(x1)

x2 = Point(2, 2)
c2 = Circle(0, 0, 2)
c2.inside(x2)


Точка (1, 1) принадлежит кругу
Точка (0, 0) принадлежит кругу
Точка (2, 2) не принадлежит кругу


### 10. Реализуйте метод класса Point, принимающий другой объект класса Point и вычисляющий эвклидово расстояние до него

In [25]:
from math import *

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        
    def evklid(self, other):
        d = sqrt(pow((self.x - other.x), 2) + pow((self.y - other.y), 2))
        print(d)

p1 = Point(10, 5)
p2 = Point(12, 7)

p1.evklid(p2)

2.8284271247461903


## ЧАСТЬ II. Решите задачи используя библиотеку NumPy

### 11. Создайте матрицу 4x4, значения элементов которой варьируются от 0 до 15

In [22]:
from numpy import *

M = arange(0,16).reshape(4, 4)

M

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

### 12. Найдите индексы ненулевых элементов из списка [1,2,0,0,4,0,5,7,8,0,0,0]

In [17]:
Z = array([1,2,0,0,4,0,5,7,8,0,0,0])
nonzero(Z)

(array([0, 1, 4, 6, 7, 8], dtype=int64),)

### 13. Создайте единичную матрицу размера 3х3

In [5]:
diag([1,1,1])

array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 1]])

### 14. Создайте матрицу 10х10, заполните её случайными значениями и найдите максимальный и минимальный элементы

In [29]:
X = random.rand(10,10)
print ('Min =', X.min(), 'Max =', X.max())

Min = 0.011639011257647658 Max = 0.9950747326661316


### 15. Создайте матрицу 5х5 и расположите в ней числа 1, 2, 3, 4 под главной диагональю

In [30]:
diag([1,2,3,4], k = -1)

array([[0, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [0, 0, 3, 0, 0],
       [0, 0, 0, 4, 0]])

### 16. Перемножьте случайные матрицы размерностей 5х3 и 3х2

In [23]:
X = random.rand(5,3)
Z = random.rand(3,2)
print(dot(X,Z))

[[0.64576699 0.83338499]
 [0.87161082 1.00427691]
 [0.52902128 0.59891736]
 [0.94133593 1.06235094]
 [0.47946194 0.50912627]]


### 17. Найдите число из списка t, наиболее близкое по величине к числу a

In [65]:
a = 2.39
t = array([0, 1, 1.2, 2, 3, 9])

index = abs(t-a).argmin()
print(t[index])

2.0


### 18. Создайте массив из 10 элементов и поменяйте значение его наибольшего элемента на 0 

In [70]:
Z = arange(10)
print(Z)
Z[Z.argmax()] = 0
print(Z)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 0]


### 19. Создайте матрицу 5x5 и поменяйте в ней местами 2 и 3 строку

In [80]:
X = random.rand(5,5)
print(X)
X[[1,2]] = X[[2,1]]
print(X)


[[0.47643784 0.20532279 0.56069596 0.93688496 0.86403841]
 [0.53033167 0.28856006 0.78175794 0.03822836 0.46162386]
 [0.86616975 0.12484369 0.7660166  0.67742296 0.97115496]
 [0.9831557  0.39708118 0.91175001 0.32359242 0.39561706]
 [0.90508371 0.69609466 0.7562698  0.25278841 0.55601309]]
[[0.47643784 0.20532279 0.56069596 0.93688496 0.86403841]
 [0.86616975 0.12484369 0.7660166  0.67742296 0.97115496]
 [0.53033167 0.28856006 0.78175794 0.03822836 0.46162386]
 [0.9831557  0.39708118 0.91175001 0.32359242 0.39561706]
 [0.90508371 0.69609466 0.7562698  0.25278841 0.55601309]]


### 20. Создайте случайную целочисленную матрицу и найдите наиболее часто встречающийся в ней элемент

In [109]:
X = random.randint(5, size=(5,5))
print(X)
Z = X.flatten()
print(bincount(Z).argmax())

[[0 2 4 3 1]
 [4 3 0 1 2]
 [1 3 4 0 4]
 [1 0 1 4 4]
 [1 1 3 0 1]]
1


## Часть III. Прочитайте краткий мануал по библиотеке pandas и выполните оставшиеся задания

Pandas - это библиотека Python, предоставляющая широкие возможности для анализа данных. С ее помощью очень удобно загружать, обрабатывать и анализировать табличные данные с помощью SQL-подобных запросов. В связке с библиотеками Matplotlib и Seaborn появляется возможность удобного визуального анализа табличных данных.


Для использования библиотек их необходимо импортировать, что можно сделать с помощью команды import. После ключевого слова import указывается название библиотеки, а после as - ее краткое название (алиас), которое мы будем использовать в коде для краткого обозначения этой библиотеки.

In [14]:
import pandas as pd

Основными структурами данных в Pandas являются классы Series и DataFrame. Первый из них представляет собой одномерный индексированный массив данных некоторого фиксированного типа (его можно интеерпретировать как индексированный столбец таблицы). Второй - это двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа. Можно представлять её как словарь объектов типа Series. Структура DataFrame отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

## Чтение из файла и первичный анализ
Чаще всего, данные, которые мы будем анализировать, готовить  и использовать для обучения моделей, не формируются самостоятельно, а берутся из готового дататсета. Рассмотрим чтение датасета (DataFrame) из файла с расширением csv. В качестве примера возьмем titanic_train.csv, который содержить данные о пассажирах Титаника. Полный датасет titanic был разделен на 2 части - тренировочную и тестовую. Именно с тренировочной частью мы и будем работать. Датасет titanic предоставлен в рабочей папке Лаб_1, а также вы можете его скачать (titanic_train.csv) по ссылке https://www.kaggle.com/tedllh/titanic-train#titanic_train.csv. Если вы работаете на платформе Goggle Colab, необходимо сделать upload файла в вашу рабочую директорию ( в меню ноутбука слева  найти и перейти на Files, нажать Upload и выбрать файл для загрузки)

Внимание! Проверьте, что файл находиться в папке content, если его там нет, то нажмите на нее правой кнопкой и upload. 


###Чтение данных

Для чтения csv файла используем функцию pd.read_csv

In [26]:
import pandas as pd
df = pd.read_csv('titanic_train.csv')

Посмотрим тип переменной df

In [27]:
type(df)

pandas.core.frame.DataFrame

### Первичный анализ

После того, как мы считали данные, посмотрим первые несколько записей датафрейма (если в скобках после функции head не указать нужное количество записей, то по умолчанию будут выведены первые 5 записей датафрейма). Если вместо таблицы вы видите  "кашу" из данных, разделенных ";", значит, ваш csv файл использует в качестве разделителя ";", а не ",". И для корректного чтения данных, нужно задать разделитель в функции чтения, например df = pd.read_csv('proba.csv', sep=";")

In [22]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Давайте посмотрим, что это за данные. Достаточно часто датасеты имеют краткое описание содержимого. Например, что означает тот или иной столбец, какие значения используются для заполнения и т.д. Или значения признаков интуитивно понятны. В нашем случае, мы имеем следующие данные:
каждая строка - это сведения об отдельном пассажире. Строка содержит:
*   PassengerId - идентификатор пассажира (проще говоря, порядковый номер строки)
*   Survived - признак, выжил ли пассажир после крушения (0 - погиб, 1 - выжил)
*   Pclass - в каком классе ехал пассажир (1, 2 или 3)
*   Name - имя пассажира
*   Sex - пол (male - мужской, female - женский)
*   Age - возраст
*   SibSp - количество родственников (по горизонтали - супруг, братья, сестры) на борту
*   Parch - количество родственников (по вертикали - родители, дети) на борту 
*   Ticket - номер билета
*   Fare - цена билета
*   Cabin - номер каюты
*   Embarked - порт посадки (C — Cherbourg, Q — Queenstown, S — Southampton)





Команда shape() показывате размерность датафрейма

Мы видим, что в нашем датафрейме 891 строка (то есть, сведения о 891 пассажире) и 12 столбцов признаков

In [23]:
df.shape

(891, 12)

Команда columns.values() показывате название всех признаков

In [24]:
df.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

Также полезно посмотреть метод info, который показывает краткую статистику по набору данных

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Можно увидеть, что не все признаки у нас заполнены в каждой строке,для некоторых признаков есть пропуски. Например, у нас 891 строка, а Age (возраст) заполнен только в 714 строках. Совсем мало значений в столбце Cabin. 

Есть несколько вариантов работы с такими строками: 
* их можно оставить без изменений, если признак, в котором пропущены данные, малозначим;
* можно удалить все строки, в которых есть пустые значения;
* можно заполнить их какими-либо значениями (0, медианой или любыми другими константами).

Метод info так же показывает какого типа данные в том или ином столбце: цифровые (int или float), текстовые и т.д. В последней строке указан объем памяти, которую занимает датафрейм. При больших объемах данных это позволяет оценить хватит ли вам ресурсов для загрузки в память полного датасета

Еще один полезный метод для начального анализа данных - метод describe(), который выдает краткую статистику по числовым признакам: среднее, стандартное отклонение, минимум, максимум, медианы, квартили и т.д. Самостоятельно посмотрите, что означает каждое понятие. 

In [26]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

##Доступ к элементам данных

###Доступ к элементам DataFrame

Разберемся как посмотреть какие-то конкретные значения в датафрейме. Наш датафрейм представляет собой таблицу, состоящую из строк и столбцов. Обратиться к конкретному столбцу можно используя его имя. Например, выведем первые 10 строк из столбца с возрастом пассажиров

In [27]:
df['Age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Второй способ обращения - через "."

In [28]:
df.Age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Столбец таблицы - это объек Pandas Series, убедимся в этом определив тип полученного объекта. Этот объект проиндексирован (слева печатаются индексы). В нижней строке есть информацию о его имени и типе значений.
Кроме того, здесь хорошо видны пропуски (незаполненные значения) в некоторых строках (они обозначены NaN - Not-A-Number)

In [29]:
type(df['Age'])

pandas.core.series.Series

Кроме обращения к столбцу по имени, можно использовать индексацию с помощью методов loc и iloc

Метод loc выдает нам подматрицу, используя индексы строк и столбцов. Первый аргумент отвечает за количество записей (как с начала, так и при желании с любого места), а второй - за отраженные в подматрице признаки.


In [30]:
df.loc[0:5, ['Pclass', 'Fare']]

Pclass     Fare
0       3   7.2500
1       1  71.2833
2       3   7.9250
3       1  53.1000
4       3   8.0500
5       3   8.4583

Метод iloc работает схожим образом, за исключением того, что второй аргумент теперь тоже индексы, то есть отражает количество признаков.


In [0]:
df.iloc[0:5, 1:3]

# Доступ к элементам Series

Для примера создадим два объекта Series

In [31]:
s1 = pd.Series(["one", "two", "three", "four", "five"]) # передаем список, индексы проставляются автоматически с 0
print(s1)

0      one
1      two
2    three
3     four
4     five
dtype: object


In [41]:
s2 = pd.Series([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e'])  # передаем список, индексы заданы символами от 'a' до 'e'
print(s2)

a    1
b    2
c    3
d    4
e    5
dtype: int64


К элементам Series можно обращаться по численному индексу через "[ ]"

In [33]:
s2[2] # второй элемент серии s2 (индексация с 0)

3

К элементам Series можно обращаться по метке:

In [0]:
s2['c']

К элементам Series можно применять так называемые slice (срезы) через ':'.
Срезами является часть серии до или после определенного элемента

In [0]:
s2[:2] # все элементы до второго включительно

In [0]:
s2[2:] # второй и все элементы после второго до конца

In [0]:
s2[2:4] # все элементы со второго по четвертый не включительно. Последний никогда не включается.

In [0]:
s2[:] # все элементы

В поле для индекса можно поместить условное выражение.
Данное выражение можно перевести как вопрос, является ли значение элемента (признака) меньшим или равным двум.

In [42]:
s2[s2 <= 4]

a    1
b    2
c    3
d    4
dtype: int64

Или даже несколько условных выражений, отделенных "( )". 

Если соединить два условия знаком "&", то результат будет удовлетворять двум условиям, а если знаком "|", то хотя бы одному из них.

In [0]:
s2[(s2 >= 2) & (s2 <=4)]

In [0]:
s2[(s2 <= 2) | (s2 == 5)]

Важно, что в условных выражениях равенство пишется как "==" (два знака "равно").

Со структурами Series можно работать как с векторами: складывать, умножать вектор на число и т.п.

In [45]:
s3 = s2 + 10

In [46]:
print(s3)

a    11
b    12
c    13
d    14
e    15
dtype: int64


## Примеры анализа информации

Для примера отберем пассажиров мужского пола, возраст которых меньше сорока лет.

In [47]:
df[(df['Sex'] == 'male') & (df.Age < 40)].head(5)

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0             1         0       3         Braund, Mr. Owen Harris  male  22.0   
4             5         0       3        Allen, Mr. William Henry  male  35.0   
7             8         0       3  Palsson, Master. Gosta Leonard  male   2.0   
12           13         0       3  Saundercock, Mr. William Henry  male  20.0   
13           14         0       3     Andersson, Mr. Anders Johan  male  39.0   

    SibSp  Parch     Ticket    Fare Cabin Embarked  
0       1      0  A/5 21171   7.250   NaN        S  
4       0      0     373450   8.050   NaN        S  
7       3      1     349909  21.075   NaN        S  
12      0      0  A/5. 2151   8.050   NaN        S  
13      1      5     347082  31.275   NaN        S

Можно отсортировать этих людей по возрастанию платы за билет. Ascending в данной функции отвечает за сортировку по убыванию (False) или по возрастанию (True).

In [48]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).head(10)

PassengerId  Survived  Pclass                             Name   Sex  \
271          272         1       3     Tornquist, Mr. William Henry  male   
822          823         0       1  Reuchlin, Jonkheer. John George  male   
806          807         0       1           Andrews, Mr. Thomas Jr  male   
179          180         0       3              Leonard, Mr. Lionel  male   
302          303         0       3  Johnson, Mr. William Cahoone Jr  male   
378          379         0       3              Betros, Mr. Tannous  male   
872          873         0       1         Carlsson, Mr. Frans Olof  male   
843          844         0       3       Lemberopolous, Mr. Peter L  male   
371          372         0       3        Wiklund, Mr. Jakob Alfred  male   
202          203         0       3       Johanson, Mr. Jakob Alfred  male   

      Age  SibSp  Parch   Ticket    Fare        Cabin Embarked  
271  25.0      0      0     LINE  0.0000          NaN        S  
822  38.0      0      0    19972  0.0000          NaN        S  
806  39.0      0      0   112050  0.0000          A36        S  
179  36.0      0      0     LINE  0.0000          NaN        S  
302  19.0      0      0     LINE  0.0000          NaN        S  
378  20.0      0      0     2648  4.0125          NaN        C  
872  33.0      0      0      695  5.0000  B51 B53 B55        S  
843  34.5      0      0     2683  6.4375          NaN        C  
371  18.0      1      0  3101267  6.4958          NaN        S  
202  34.0      0      0  3101264  6.4958          NaN        S

Можно подсчитать количество пассажиров (а точнее их уникальных номеров) с помощью функции count()

In [49]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).PassengerId.count()

344

Также можно быстро вывести распределение переменной Sex с помощью функции value_count()

In [50]:
df.Sex.value_counts()

male      577
female    314
Name: Sex, dtype: int64

## Задания

### 21. Сколько человек спаслись и сколько погибли?

In [33]:
df.Survived.value_counts()

0    549
1    342
Name: Survived, dtype: int64

### 22. Выведите распределение переменной Survived отдельно для мужчин и отдельно для женщин


In [58]:
SurvMale = df[(df['Sex'] == 'male')].Survived.value_counts()
SurvFemale = df[(df['Sex'] == 'female')].Survived.value_counts()
print('Survived or Died for Male:\n', SurvMale)
print('Survived or Died for Female:\n', SurvFemale)

Survived or Died for Male:
 0    468
1    109
Name: Survived, dtype: int64
Survived or Died for Female:
 1    233
0     81
Name: Survived, dtype: int64


### 23. Каковы медиана и стандартное отклонение цен за билет?

In [61]:
print('median:', df.Fare.median())
print('std:', df.Fare.std())

median: 14.4542
std: 49.693428597180905


### 24. Подсчитайте долю выживших среди людей моложе 25 лет и старше 55.

In [28]:
df = pd.read_csv('titanic_train.csv')
df_new = df[(df['Age'] < 25) | (df.Age > 55)]
surv_count = df_new['Survived'].value_counts()
x = surv_count[1] / surv_count.sum()
print(x)

0.4088050314465409


### 25. Подсчитайте долю выживших среди мужчин и женщин.

In [91]:
df_male = df[(df['Sex'] == 'male')]
surv_count_male = df_male['Survived'].value_counts()
rate_male = surv_count_male[1] / surv_count.sum()


df_female = df[(df['Sex'] == 'female')]
surv_count_female = df_female['Survived'].value_counts()
rate_female = surv_count_female[1] / surv_count.sum()

print('Male survival rate:', rate_male)
print('Female survival rate:', rate_female)

Male survival rate: 0.34276729559748426
Female survival rate: 0.7327044025157232


### 26. В каком городе больше всего село пассажиров?

In [31]:
df_city = df['Embarked'].value_counts()
df_city.argmax()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  


'S'

### 27. Подсчитайте средний возраст выживших

In [101]:
mean_age = df[(df['Survived'] == 1)].Age.mean()
print(mean_age)

28.343689655172415


### 28. Подсчитайте средний возраст пассажиров каждого из трех классов


In [103]:
mean_fc = df[(df['Pclass'] == 1)].Age.mean()
mean_sc = df[(df['Pclass'] == 2)].Age.mean()
mean_tc = df[(df['Pclass'] == 3)].Age.mean()
print('Mean first class', mean_fc)
print('Mean second class', mean_sc)
print('Mean third class', mean_tc)

Mean first class 38.233440860215055
Mean second class 29.87763005780347
Mean third class 25.14061971830986


### 29. Подсчитайте средний возраст выживших мужчин, купивших билет во 2 класс

In [109]:
mean_male = df[(df['Pclass'] == 2) & (df.Sex == 'male') & (df.Survived == 1)].Age.mean()
print(mean_male)

16.022


### 30. Оцените, как влияет на выживаемость пассажиров наличие их родственников на борту.

In [113]:
with_relation = df[(df.SibSp != 0) & (df.Parch != 0)].Survived.value_counts()
with_relation_ratio = with_relation[1] / with_relation.sum()

without_relation = df[(df.SibSp == 0) & (df.Parch == 0)].Survived.value_counts()
without_relation_ratio = without_relation[1] / without_relation.sum()

print('Survival rate with relations: ', with_relation_ratio)
print('Survival rate without relations: ', without_relation_ratio)

Survival rate with relations:  0.43661971830985913
Survival rate without relations:  0.30353817504655495
